### DISCUSSION TITLE
Create `app.timer` for timing UI-independent functions

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/nicegui#3225] [Ideas / Feature Requests]
As discussed in #3187, `ui.timer` works like a UI element: It starts running when a page is created and stops when removing the timer or its parent container. This is handy in some cases, but isn't very helpful in others.

Therefore we propose an `app.timer` function that behaves similar to `ui.timer`, but runs on an app level, independent of connected clients. It will, however, have no UI context to work with. So it is primarily meant to run non-UI code.

**Comment by v479038280** [comment]
any progress？

**Reply by falkoschindler** [reply]
Good question!
I took the opportunity to create PR #4091 implementing this feature, borrowing some code from [RoSys](https://github.com/zauberzeug/rosys).
There shouldn't be too much work left, so we might be able to release it in upcoming version 2.9.

**Comment by thetableman** [comment]
I know I'm a bit late to the party, given that a PR is already in the works, but I've been using a variant of this package along with an `@app.on_startup(...)` to automate a regular background process.

https://fastapi-utils.davidmontague.xyz/user-guide/repeated-tasks/#the-repeat_every-decorator

I do like the idea of the timer being implemented as a decorator function versus how I expect the version in the PR will work.

**Reply by falkoschindler** [reply]
That's a great idea, @thetableman!
I don't think a decorator contradicts PR #4091. It could be an extension so that you can either use `app.timer` like `ui.timer` (in fact one is deriving from another) or `@app.repeat_every(...)`. If `repeat_every` would use `app.timer` internally, it would automatically run on startup (or immediately if the app is already running).

**Reply by falkoschindler** [reply]
Simply adding
```py
def repeat_every(self, seconds: float) -> Callable:
    """Decorator to repeat a function every `seconds` seconds."""
    def decorator(func: Callable) -> Callable:
        self.timer(seconds, func)
        return func
    return decorator
```
to app.py seems to work.

**Reply by rodja** [reply]
Should we add this feature in another pull request?

**Comment by thetableman** [comment]
Another thought, perhaps for a later version, would be to extend this proposed app level timer capabilities to run functions at specific times. 

A simpler implementation might allow a user to just schedule a function at 6.00am each day whilst a more advanced version might allow for every Tuesday at 6.00am or the 15th of every Month.

**Reply by falkoschindler** [reply]
That would be nice, but maybe out of scope of a UI framework. You could use packages like [schedule](https://schedule.readthedocs.io/en/stable/) for this purpose.

**Reply by thetableman** [reply]
True... but `schedule` also supports running a function at a timed interval (`schedule.every(10).minutes.do(job)`) so wouldn't that then negate the need for developing a custom `app.timer` function?

**Reply by falkoschindler** [reply]
Valid point. In our applications, especially our robot systems, we often need to run callbacks every x seconds to read sensors, update internal state etc. So it is very convenient to have such timers built into NiceGUI - like setTimeout and setInterval in JavaScript. For scheduling callbacks for a certain time of day, we created an application-specific [schedule](https://github.com/zauberzeug/rosys/blob/main/rosys/automation/schedule.py) module which also takes sunrise and sunset times into account. Requirements can vary between applications, which is why we didn't move it to NiceGUI.
What I'm trying to say: Looping or delaying with a simple interval given in seconds is a relatively straight-forward requirement covering the majority of use cases. But we'll consider adding some kind of simple schedule feature in the future. 👍

**Comment by falkoschindler** [comment]
`app.timer` has just been released in NiceGUI 2.9. 🚀

Let's discuss ideas like `@app.repeat_every` and scheduling function calls on separate feature requests.
